pkg/queue/queue_test.go

```diff
@@ -333,6 +336,208 @@ func TestMaxQueueSize(t *testing.T) {
+			tasksPerTenant:   10,
+			consumersCount:   1,
+			dequeueBatchSize: 2,
+		},
+		"tenants-1_tasks-10_consumers-1_batch-2": {
+			tenantsCount:     1,
+			tasksPerTenant:   10,
+			consumersCount:   1,
+			dequeueBatchSize: 2,
+		},
+		"tenants-1_tasks-10_consumers-10_batch-2": {
+			tenantsCount:     1,
+			tasksPerTenant:   10,
+			consumersCount:   10,
+			dequeueBatchSize: 2,
+		},
+		"tenants-100_tasks-10_consumers-10_batch-6_maxConsumersPerTenant-6": {
+			tenantsCount:          100,
+			tasksPerTenant:        10,
+			consumersCount:        10,
+			dequeueBatchSize:      6,
+			maxConsumersPerTenant: 4,
+		},
+		"tenants-200_tasks-10_consumers-100_batch-999999_maxConsumersPerTenant-4": {
+			tenantsCount:          200,
+			tasksPerTenant:        10,
+			consumersCount:        100,
+			dequeueBatchSize:      999_999,
+			maxConsumersPerTenant: 4,
+		},
+		"tenants-10_tasks-100_consumers-20_batch-5_maxConsumersPerTenant-16": {
+			tenantsCount:          10,
+			tasksPerTenant:        100,
+			consumersCount:        20,
+			dequeueBatchSize:      5,
+			maxConsumersPerTenant: 16,
+		},
+		"tenants-1_tasks-100_consumers-16_batch-5_maxConsumersPerTenant-2": {
+			tenantsCount:          1,
+			tasksPerTenant:        100,
+			consumersCount:        16,
+			dequeueBatchSize:      5,
+			maxConsumersPerTenant: 2,
+		},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			totalTasksCount := tt.tenantsCount * tt.tasksPerTenant
+			limits := &mockLimits{maxConsumer: tt.maxConsumersPerTenant}
+			require.LessOrEqual(t, tt.tasksPerTenant, tenantsQueueMaxSize, "test must be able to enqueue all the tasks without error")
+			queue := NewRequestQueue(tenantsQueueMaxSize, 0, limits, NewMetrics(nil, constants.Loki, "query_scheduler"))
+			for i := 0; i < tt.consumersCount; i++ {
+				queue.RegisterConsumerConnection(createConsumerName(i))
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+			wg, _ := errgroup.WithContext(ctx)
+
+			mtx := &sync.Mutex{}
+			receivedTasksPerTenant := make(map[string]int, tt.tenantsCount)
+			receivedTasksCount := atomic.NewInt32(0)
+			for i := 0; i < tt.consumersCount; i++ {
+				consumer := createConsumerName(i)
+				wg.Go(func() error {
+					idx := StartIndexWithLocalQueue
+					for {
+						tasks, newIdx, err := queue.DequeueMany(ctx, idx, consumer, tt.dequeueBatchSize)
+						receivedTasksCount.Add(int32(len(tasks)))
+
+						if err != nil && !errors.Is(err, context.Canceled) {
+							return fmt.Errorf("error while dequeueing many task by %s: %w", consumer, err)
+						}
+						if err == nil && len(tasks) == 0 {
+							return fmt.Errorf("%s must receive at least one item if there is no error, but got 0", consumer)
+						}
+						if len(tasks) > 1 && !isAllItemsSame(tasks) {
+							return fmt.Errorf("expected all items to be from the same tenant, but they were not: %v", tasks)
+						}
+						time.Sleep(tt.consumerDelay)
+						mtx.Lock()
```

Comment:
For better readability, maybe defer `mtx.Unlock()` here, instead of multiple times on different code branches.

Reply:
done